### PR TITLE
Add the possibility to skip light update and fix light later

### PR DESCRIPTION
**Pull request by juhdanad**
For the request of @HybridDog this patch adds the possibility to skip light update when writing voxel manipulators to the map.

`write_to_map([do_light_update])` got a boolean parameter: if it is false, light is not checked. You might want to do this if your voxel manipulator is huge and you want to split the light calculation into smaller parts, or if you modify the same area twice (like the TNT mod in MTG) and you want to skip the first, unnecessary light calculation.

If you want to fix the lighting later (or you find a lighting bug which you want to correct), call `minetest.fix_light(pos)`. This method corrects the lighting in one map block (without copying the data into a new voxel manipulator).

### CODE REVIEW

**Comment by juhdanad** [comment]
Help for reviewers:

1. the first commit only moves code.
2. the second commit mostly copies methods from the voxel manipulator light code. This is necessary, as voxel manipulators and map blocks are similar, but they need to be accessed differently.
3. in the third commit I expose `fix_light()` to Lua and add the possibility to disable light updates.

**Comment by SmallJoker** [comment]
This is a quite big change for an additional argument and command. But if it works well this way, I'm fine with it.

**Comment by HybridDog** [comment]
thanks for your efforts.
If l set the do_light_update parameter to false, the affected nodes either have the previous light (if the previous and the new node have paramtype "light") or 0, don't they?

**Comment by juhdanad** [comment]
@HybridDog if you disable light update, then `write_to_map` behaves exactly the same way as it did before the bulk light bugfix. This means that `param1` is written back to the map without modifications and nodes will have the same light levels as before (unless you modified light levels manually in the voxel manipulator).

**Comment by paramat** [comment]
> You might want to do this if your voxel manipulator is huge and you want to split the light calculation into smaller parts,

I'm unsure this is a good reason for the feature, but it may be useful for other reasons.

**Comment by juhdanad** [comment]
@paramat 

> The mapfix feature is of course useful, but could be implemented without the above feature.

The mapfix feature is the 99% of the changeset, the remaining 1% is the boolean in `write_to_map`.

@HybridDog could you update your worldedit fork to use this API? Then reviewers could test this pull with your mod.

**Comment by HybridDog** [comment]
> could you update your worldedit fork to use this API? Then reviewers could test this pull with your mod.

ok.
#
When l set a 100^3 cobble chunk with original worldedit, it now takes ~0.35 s (~5.78 s before, 16 times slower).
l noticed a strange behaviour: When replacing the cobble chunk with stone it takes ~0.41 s every time and when replacing the stone chunk with cobble it always takes ~0.32 s.
#
When deferring the light updates, it takes ~0.15 s (~0.05 s before, 3 times faster).
https://github.com/HybridDog/WorldEdit-pack/tree/bl_def

**Comment by paramat** [comment]
Striclty speaking your mod doesn't make the light updates faster, it just spreads them out, the total time updating light is probably longer. Spreading out the light updates is clever but a little messy.
I feel this PR should be considered not for 1 person's mod but for other good reasons.

**Comment by paramat** [comment]
Your Worldedit mod doesn't make the light update faster, it splits it up and spreads these out in time, the total time is probably longer.

**Comment by HybridDog** [comment]
Before bulk lighting fix the total time was 100 times shorter if you simply rejoined.

Do light bugs appear if minetest.fix_light isn't executed, e.g. because of rejoining?

**Comment by juhdanad** [comment]
What is 'rejoining'?

**Comment by HybridDog** [comment]
By rejoining l mean restarting the singleplayer game.

**Comment by juhdanad** [comment]
@HybridDog so you modify a large part of the map, and quit before lighting is updated.

If you defer light updates, you should call fix_light on each map block you modified because they may contain invalid lighting. If not all blocks are fixed, then unfixed blocks may contain lighting bugs (exception: when you replace air with opaque blocks it is enough to fix the borders to remove light that came from the air blocks)

Fix_light does exactly the same thing as writing back and updating a voxel manipulator did before the bulk light change.

File: src/script/lua_api/l_env.cpp [hunk 1]
```diff
@@ -832,6 +832,29 @@ int ModApiEnvMod::l_line_of_sight(lua_State *L)
...
 
+// fix_light(p)
+int ModApiEnvMod::l_fix_light(lua_State *L)
+{
+	GET_ENV_PTR;
+
+	v3s16 blockpos = read_v3s16(L, 1);
+	ServerMap &map = env->getServerMap();
+	std::map<v3s16, MapBlock *> modified_blocks;
+	bool success = map.repairBlockLight(blockpos, &modified_blocks);
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: this variable is not needed, call it in condition

**Comment by HybridDog** [comment]
l noticed a severe bug: the map is not written everywhere. l'm not sure whether this is caused by this PR, bulk lighthing revision or something else.
screenshot_20170316_211505

**Comment by paramat** [comment]
> the map is not written everywhere.

How did you cause this bug? It may be due to your use of worledit, too large a volume set?

**Comment by HybridDog** [comment]
l found a way to reproduce this with usual worldedit:
* //teleport 1000 0 1000
⇒ map there becomes generated
* //pos1
* //teleport 1000 1000 1000
⇒ there map is generated too
* //pos2
* //set wood
* //set cobble
* rejoin (restart) now
* explore the not generated map

The bug appears in not generated mapblocks. The 1x1 pillar correctly consists of cobble at the ends, where the map is generated. The other parts of the pillar, after restarting the server, consist of wood, and sometimes air.
Without restarting or in already generated map the bug doesn't appear.

**Comment by HybridDog** [comment]
l moved it to an issue: #5408.

**Comment by HybridDog** [comment]
l just noticed that you currently can't change the light manually, the set_light_data method only changes param1 of nodes which don't have paramtype "light".
For my shadows mod l need to manually change the light values.
Could you automatically set do_light_update to false when the set_light_data method is used?

**Comment by juhdanad** [comment]
@HybridDog

> Could you automatically set do_light_update to false when the set_light_data method is used?

That would make the API more complicated, I think the current implementation is better.

File: src/script/lua_api/l_env.cpp [hunk 2]
```diff
@@ -832,6 +832,29 @@ int ModApiEnvMod::l_line_of_sight(lua_State *L)
...
+	if (success) {
+		MapEditEvent event;
+		event.type = MEET_OTHER;
+		for (std::map<v3s16, MapBlock *>::iterator it = modified_blocks.begin();
+			it != modified_blocks.end(); ++it)
```
Comment (SmallJoker) [inline src/script/lua_api/l_env.cpp]: One indent missing

File: src/voxelalgorithms.cpp [hunk 3]
```diff
@@ -1094,6 +1094,95 @@ const VoxelArea block_pad[] = {
...
+			*modified_blocks);
+	}
+
+	// --- STEP 2: Get all newly inserted light sources
+
+	// For each block:
+	for (s16 b_x = minblock.X; b_x <= maxblock.X; b_x++)
+	for (s16 b_y = minblock.Y; b_y <= maxblock.Y; b_y++)
+	for (s16 b_z = minblock.Z; b_z <= maxblock.Z; b_z++) {
+		v3s16 blockpos(b_x, b_y, b_z);
```
Comment (SmallJoker) [inline src/voxelalgorithms.cpp]: Const here too?

File: src/voxelalgorithms.cpp [hunk 4]
```diff
@@ -1187,30 +1276,105 @@ void blit_back_with_light(ServerMap *map, MMVManip *vm,
...
+		// For each node, downwards:
+		for (s16 y = MAP_BLOCKSIZE - 1; y >= 0; y--) {
+			MapNode n = block->getNodeNoCheck(x, y, z, &is_valid);
+			// Ignore IGNORE nodes, these are not generated yet.
+			if (n.getContent() == CONTENT_IGNORE)
+				continue;
+			const ContentFeatures &f = ndef->get(n.getContent());
+			if (lig && !f.sunlight_propagates)
+				// Sunlight is stopped.
+				lig = false;
```
Comment (SmallJoker) [inline src/voxelalgorithms.cpp]: This is valid but you should put brackets around this block anyway to prevent misunderstandings.

**Comment by juhdanad** [comment]
The code's style is fixed!

**Comment by paramat** [comment]
Seems useful.(for reasons other than multiple light updates in a large vm).

**Comment by juhdanad** [comment]
I added one more commit, which changes the parameters from block coordinates to node coordinates. The new syntax is: `minetest.fix_light(pos1, pos2)`

**Comment by paramat** [comment]
Perhaps the documentation can explain that all mapblocks are processed that intersect with yhe specified volume? Many modders do know what mapblocks are and need to know exactly what volume will be processed. Current docs do not explain how much larger than the specified volume the processed volume will be.

**Comment by paramat** [comment]
@HybridDog it's not a good idea to use the official approve function, as that is for devs to shoe official approval.

**Comment by HybridDog** [comment]
> Please remove it.

l can't.

**Comment by juhdanad** [comment]
@paramat Now the documentation describes exactly which area is changed.

**Comment by paramat** [comment]
Thanks seems good.

HybridDog no problem, thanks for trying.

File: src/script/lua_api/l_env.cpp [hunk 5]
```diff
@@ -832,6 +832,39 @@ int ModApiEnvMod::l_line_of_sight(lua_State *L)
...
+	bool success = true;
+	v3s16 blockpos;
+	for (blockpos.X = blockpos1.X; blockpos.X <= blockpos2.X;
+			blockpos.X++)
+		for (blockpos.Y = blockpos1.Y; blockpos.Y <= blockpos2.Y;
+				blockpos.Y++)
+			for (blockpos.Z = blockpos1.Z; blockpos.Z <= blockpos2.Z;
+					blockpos.Z++)
+				success = success
+					& map.repairBlockLight(blockpos, &modified_blocks);
```
Comment (paramat) [inline src/script/lua_api/l_env.cpp]: For clarity and compactness
```
	for (blockpos.X = blockpos1.X; blockpos.X <= blockpos2.X; blockpos.X++)
	for (blockpos.Y = blockpos1.Y; blockpos.Y <= blockpos2.Y; blockpos.Y++)
	for (blockpos.Z = blockpos1.Z; blockpos.Z <= blockpos2.Z; blockpos.Z++) {
		success = success & map.repairBlockLight(blockpos, &modified_blocks);
	}
```

File: src/voxelalgorithms.cpp [hunk 6]
```diff
@@ -1094,6 +1094,95 @@ const VoxelArea block_pad[] = {
...
+			v3s16 relpos(x, y, z);
+			MapNode node = block->getNodeNoCheck(x, y, z, &is_valid);
+			const ContentFeatures &f = ndef->get(node);
+			// For each light bank
+			for (size_t b = 0; b < 2; b++) {
+				LightBank bank = banks[b];
+				u8 light = f.param_type == CPT_LIGHT ?
+					node.getLightNoChecks(bank, &f):
+					f.light_source;
+				if (light > 1)
```
Comment (paramat) [inline src/voxelalgorithms.cpp]: Just checking: does light = 1 need relight?

File: src/voxelalgorithms.cpp [hunk 7]
```diff
@@ -1187,30 +1276,106 @@ void blit_back_with_light(ServerMap *map, MMVManip *vm,
...
+	// Extract sunlight above.
+	is_sunlight_above_block(map, blockpos, ndef, lights);
+	// Reset the voxel manipulator.
+	fill_with_sunlight(block, ndef, lights);
+	// Copy sunlight data
+	data.target_block = v3s16(blockpos.X, blockpos.Y - 1, blockpos.Z);
+	for (s16 z = 0; z < MAP_BLOCKSIZE; z++)
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+		data.data.push_back(
+			SunlightPropagationUnit(v2s16(x, z), lights[z][x]));
```
Comment (paramat) [inline src/voxelalgorithms.cpp]: For clarity
```
	for (s16 z = 0; z < MAP_BLOCKSIZE; z++)
	for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
		data.data.push_back(SunlightPropagationUnit(v2s16(x, z), lights[z][x]));
	}

**Comment by paramat** [comment]
Line comments.
I only partly understand the code.

**Comment by paramat** [comment]
:+1: Tested and works.

**Comment by paramat** [comment]
It would be good to have a /fixlight <here (radius)> <pos1, pos2> chat command in a separate PR. There are already helper functions for this in chatcommands.lua, it could work similar to /deleteblocks.

**Comment by Zeno-** [comment]
:+1:

**Comment by Ekdohibs** [comment]
https://github.com/minetest/minetest/commit/6d1e6f889826a5802e17f53f99000a51b2e00066
https://github.com/minetest/minetest/commit/57e5aa662851485902575c3c747437e365bf72c8

File: src/voxelalgorithms.cpp (hunk 6 above)
Comment (juhdanad) [inline src/voxelalgorithms.cpp]: No, because light=1 won't make its neighbors brighter.

**Comment by paramat** [comment]
#5630 merged.

**Pull request closed without merging**